osh/word_eval.py

```diff
@@ -684,14 +685,45 @@ def _ApplyTestOp(
                 else:
                     is_falsey = False
 
-            elif case(value_e.BashArray):
-                val = cast(value.BashArray, UP_val)
-                # TODO: allow undefined
-                is_falsey = len(val.strs) == 0
+            elif case(value_e.BashArray, value_e.BashAssoc):
+                if val.tag() == value_e.BashArray:
+                    val = cast(value.BashArray, UP_val)
+                    strs = bash_impl.BashArray_GetValues(val)
+                elif val.tag() == value_e.BashAssoc:
+                    val = cast(value.BashAssoc, UP_val)
+                    strs = bash_impl.BashAssoc_GetValues(val)
+                else:
+                    raise AssertionError()
 
-            elif case(value_e.BashAssoc):
-                val = cast(value.BashAssoc, UP_val)
-                is_falsey = len(val.d) == 0
+                if tok.id in (Id.VTest_ColonHyphen, Id.VTest_ColonEquals,
+                              Id.VTest_ColonQMark, Id.VTest_ColonPlus):
+                    # The first character of IFS is used as a separator only
+                    # for the double-quoted "$*", or otherwise, a space " " is
+                    # used (for $*, $@, and "$@").
+                    # TODO: We current do not check whether the current $* is
```

Comment:
Hm I noticed this function has a "quoted" param, we might be able to use that

Reply:
Oh, OK. I haven't noticed that it corresponds to the condition I wanted!